core/DataAccess/Model.php

```diff
@@ -440,4 +539,161 @@ private static function getNameCondition($doneFlags, $possibleValues)
         return $result;
     }
 
+    /**
+     * Marks an archive as in progress if it has not been already. This method must be thread
+     * safe.
+     */
+    public function startArchive($idSite, $date1, $date2, $period, $doneFlag)
+    {
+        $table = ArchiveTableCreator::getNumericTable(Date::factory($date1));
+
+        // find latest idarchive
+        $sql = "SELECT ts_archived, idarchive, `value`
+            FROM `$table`
+            WHERE idsite = ? AND date1 = ? AND date2 = ? AND period = ? AND `name` = ?
+            ORDER BY ts_archived DESC
+            LIMIT 1";
+        $bind = [$idSite, $date1, $date2, $period, $doneFlag];
+        $latestArchive = Db::fetchRow($sql, $bind);
+
+        if (empty($latestArchive)) { // should never happen
+            return null;
+        }
+
+        // if the archive is done or being processed, we don't need to do anything so we abort
+        if (!empty($latestArchive)
+            && ($latestArchive['value'] == ArchiveWriter::DONE_OK
+                || $latestArchive['value'] == ArchiveWriter::DONE_IN_PROGRESS)
+        ) {
+            return null;
+        }
+
+        // set archive value to DONE_IN_PROGRESS IF NOT SET ALREADY
+        $statement = Db::query("UPDATE `$table` SET `value` = ? WHERE idarchive = ? AND `name` = ? AND value = ?", [
+            ArchiveWriter::DONE_IN_PROGRESS,
+            $latestArchive['idarchive'],
+            $doneFlag,
+            ArchiveWriter::DONE_INVALIDATED,
+        ]);
+
+        if ($statement->rowCount() > 0) { // if we updated, then we've marked the archive as started
+            return $latestArchive['idarchive'];
+        }
+
+        // if we didn't get anything, some process either got there first, OR
+        // the archive was started previously and failed in a way that kept it's done value
+        // set to DONE_IN_PROGRESS. try to acquire the lock and if acquired, archiving isn' in process
+        // so we can claim it.
+        $lock = $this->archivingStatus->acquireArchiveInProgressLock($idSite, $date1, $date2, $period, $doneFlag);
+        if (!$lock->isLocked()) {
+            return null; // we couldn't claim the lock, archive is in progress
+        }
+
+        Db::query("UPDATE `$table` SET `value` = ? WHERE idarchive = ? AND `name` = ?", [
+            ArchiveWriter::DONE_IN_PROGRESS, $latestArchive['idarchive'], $doneFlag]);
+
+        return $latestArchive['idarchive'];
+    }
+
+    /**
+     * Gets the next invalidated archive that should be archived in a table.
+     *
+     * @param string[] $tables
+     * @param int $count
+     */
+    public function getNextInvalidatedArchive($table, $period = null, $idSites = null, $idArchivesToExclude = null)
+    {
+        $sql = "SELECT idarchive, idsite, date1, date2, period, `name`
+                  FROM `$table`
+                 WHERE `name` LIKE 'done%' AND `value` = ?";
```

Comment:
Might need some new index on `name`  covering the first 4 characters? that might help? could maybe also include value not sure... ` add index (name(4), value)`... I suppose could also first not have the index, then add `name` index should it be needed, and then `name,value`. I reckon the `name(4)` index might be needed. Of course making writes slower... so will be a tradeoff which needs to be seen. Maybe lets leave it for now without?

Reply:
> Also wondering: could it be possible that two concurrently running archivers select the same next invalidated archive?

It's possible though I tried to prevent it. The first pass just picks the invalidated archives. Later before we launch archiving we update the value to DONE_IN_PROGRESS if the value is not already set to that. That should block two archivers from being able to start a single archive, correct?

> btw this query will cause a full table scan and can be very slow.

Yeah, that's one of the reasons I brought up refactoring the archive table structure in slack. It's annoying we store both data and archive status together. I'm not too sure how to optimize this query w/ the current tables.

Reply:
Maybe we can run this query on a large table w/ instead `LIKE 'abcd%' LIMIT 1`? Then we can see what the worst case will be (if we have to do a full table scan). Or I guess w/o the LIMIT.